### ISSUE TITLE
Conv2D layer incompatible with float64 dtype

### ISSUE THREAD

**Issue by joaocaldeira** [keras-team/keras#10964]
I am using keras with tensorflow backend, and the current implementation of Conv2D layers seems incompatible with trying to use a larger floating point precision. Here is a short script that will reproduce the problem:

```
from keras.layers import Input, Conv2D
from keras import backend as K
K.set_floatx('float64')

x = Input(shape=(128,128,1))
x = Conv2D(1,3)(x)
```

Leads to `TypeError: Input 'filter' of 'Conv2D' Op has type float64 that does not match type float32 of argument 'input'`

Going to https://github.com/keras-team/keras/blob/master/keras/backend/tensorflow_backend.py#L3424 reveals the cause, the following snippet of code applied to inputs of convolutional layers:

```
if dtype(x) == 'float64':
    x = tf.cast(x, 'float32')
```

I suppose I could just remove this from keras and try again, but before going that way it seems I should ask why this is here. Is there a way to switch keras to use higher precision in conv layers, or should I drop keras and deal with tensorflow directly?

**Comment by joaocaldeira** [comment]
Nevermind, I should have looked at closed issues as well before opening this one. I see it's a problem with tensorflow. See https://github.com/tensorflow/tensorflow/pull/12943

**Comment by joaocaldeira** [comment]
Ok, I take it back (and apologize for the conversation with myself): as of tf v1.10, support for `float64` in `conv2d` has been added. https://github.com/tensorflow/tensorflow/commit/e3468b56d323783fdfb79fa2d6c24effc58bcaa9

What needs to be changed in keras to enable this as well?

**Comment by bchu** [comment]
The problem is in `_preprocess_conv1d_input`, `_preprocess_conv2d_input`, and `_preprocess_conv3d_input` in `tensorflow_backend.py`:

```
if dtype(x) == 'float64':
    x = tf.cast(x, 'float32')
```

**Comment by joaocaldeira** [comment]
Solved in #10977
